### ISSUE TITLE
Sly only works with sly connect

### ISSUE THREAD

**Issue by anquegi** [joaotavora/sly#124]
I cannot make sly working with the command sly and I must always use sly-connect

This are my settings, I'm using roswell and OSX, and on emacs 25.2:

```
;;; my-sly.el --- init emacs configuration -*- lexical-binding: t; -*-
;;; Commentary:
;;; my sly confiuration file
;;; Code:

;; regular SLY setup

(require 'sly-autoloads)

(setq inferior-lisp-program "ros dynamic-space-size=3gb -Q -- run")

(require 'sly-quicklisp-autoloads)

(electric-pair-mode +1)

(require 'rainbow-delimiters)

(add-hook 'sly-repl-mode-hook 'rainbow-delimiters-mode)
(add-hook 'sly-mode-hook 'rainbow-delimiters-mode)

(add-hook 'sly-mode-hook 'sly-company-mode)
(add-to-list 'company-backends 'sly-company)

(add-to-list 'auto-mode-alist '("\\.lass\\'" . sly-mode))


(provide 'my-sly)

;;; my-sly.el ends here
```

using sly-connect I prepare the lisp connection using tmux:

```
─ ~/.tmuxinator  2.4.0@learn  SBCL 1.3.21 
╰─ cat sly.yml                                                                                                                          ✓  7607  08:24:30
# ~/.tmuxinator/sketch.yml

name: sly
root: ~/learn/lisp

# Optional tmux socket
# socket_name: foo

# Runs before everything. Use it to start daemons etc.
# pre: ros run -L ccl/1.11

windows:
  - sbcl:
      - ros dynamic-space-size=3gb -Q run -- --load ~/.tmuxinator/create-sly.lisp
╭─ ~/.tmuxinator  2.4.0@learn  SBCL 1.3.21 
╰─ cat create-sly.lisp                                                                                                                  ✓  7608  08:24:39
(ql:quickload :swank)
(ql:quickload :slynk)
(slynk:create-server :port 4008)
```
this works pretty well: sly, sly-company and sly-quicklisp work together and really well, I feel it faster and better than slime, is rally agood IDE for common lisp, unsin M-x sly-connect

<img width="1280" alt="captura de pantalla 2017-09-09 a las 8 23 17" src="https://user-images.githubusercontent.com/2518346/30237719-2d6a2b5e-9538-11e7-9a44-f960b5bde0d2.png">

but using M-x sly I cannot make it working:

I only get a list of infinite pollings:

```
[sly] SLIME detected in `lisp-mode-hook', which causes keybinding conflicts.
Remove it for this Emacs session? (y or n) y
[sly] Polling "/tmp/sly.4932" .. 1 (Abort with `M-x sly-abort-connection'.)
[sly] Polling "/tmp/sly.4932" .. 2 (Abort with `M-x sly-abort-connection'.)
[sly] Polling "/tmp/sly.4932" .. 3 (Abort with `M-x sly-abort-connection'.)
[sly] Polling "/tmp/sly.4932" .. 4 (Abort with `M-x sly-abort-connection'.)
[sly] Polling "/tmp/sly.4932" .. 5 (Abort with `M-x sly-abort-connection'.)
[sly] Polling "/tmp/sly.4932" .. 6 (Abort with `M-x sly-abort-connection'.)
[sly] Polling "/tmp/sly.4932" .. 7 (Abort with `M-x sly-abort-connection'.)
[sly] Polling "/tmp/sly.4932" .. 8 (Abort with `M-x sly-abort-connection'.)
[sly] Polling "/tmp/sly.4932" .. 9 (Abort with `M-x sly-abort-connection'.)
[sly] Polling "/tmp/sly.4932" .. 10 (Abort with `M-x sly-abort-connection'.)
[sly] Polling "/tmp/sly.4932" .. 11 (Abort with `M-x sly-abort-connection'.)
[sly] Polling "/tmp/sly.4932" .. 12 (Abort with `M-x sly-abort-connection'.)
[sly] Polling "/tmp/sly.4932" .. 13 (Abort with `M-x sly-abort-connection'.)
[sly] Polling "/tmp/sly.4932" .. 14 (Abort with `M-x sly-abort-connection'.)
[sly] Polling "/tmp/sly.4932" .. 15 (Abort with `M-x sly-abort-connection'.)
[sly] Polling "/tmp/sly.4932" .. 16 (Abort with `M-x sly-abort-connection'.)
[sly] Polling "/tmp/sly.4932" .. 17 (Abort with `M-x sly-abort-connection'.)
[sly] Polling "/tmp/sly.4932" .. 18 (Abort with `M-x sly-abort-connection'.)
[sly] Polling "/tmp/sly.4932" .. 19 (Abort with `M-x sly-abort-connection'.)
[sly] Polling "/tmp/sly.4932" .. 20 (Abort with `M-x sly-abort-connection'.)
[sly] Polling "/tmp/sly.4932" .. 21 (Abort with `M-x sly-abort-connection'.)
[sly] Polling "/tmp/sly.4932" .. 22 (Abort with `M-x sly-abort-connection'.)
[sly] Polling "/tmp/sly.4932" .. 32 (Abort with `M-x sly-abort-connection'.)
[sly] Polling "/tmp/sly.4932" .. 33 (Abort with `M-x sly-abort-connection'.)
[sly] Polling "/tmp/sly.4932" .. 34 (Abort with `M-x sly-abort-connection'.)
[sly] Polling "/tmp/sly.4932" .. 35 (Abort with `M-x sly-abort-connection'.)
[sly] Polling "/tmp/sly.4932" .. 36 (Abort with `M-x sly-abort-connection'.)
[sly] Polling "/tmp/sly.4932" .. 37 (Abort with `M-x sly-abort-connection'.)
[sly] Polling "/tmp/sly.4932" .. 38 (Abort with `M-x sly-abort-connection'.)
[sly] Polling "/tmp/sly.4932" .. 39 (Abort with `M-x sly-abort-connection'.)
[sly] Polling "/tmp/sly.4932" .. 40 (Abort with `M-x sly-abort-connection'.)
[sly] Polling "/tmp/sly.4932" .. 41 (Abort with `M-x sly-abort-connection'.)
[sly] Polling "/tmp/sly.4932" .. 42 (Abort with `M-x sly-abort-connection'.)
[sly] Polling "/tmp/sly.4932" .. 43 (Abort with `M-x sly-abort-connection'.)
[sly] Polling "/tmp/sly.4932" .. 44 (Abort with `M-x sly-abort-connection'.)
... and continues forever
```

**Comment by joaotavora** [comment]
I supposes it's some unexpected way in which the:

```
ros dynamic-space-size=3gb -Q -- run
```

that you set for `inferior-lisp-mode` answers back to emacs. But I need more output. Can you switch, during the infinite polling, to the buffer called something like `*sly-started inferior-lisp for ros...`?

Also, can you try with something simpler like

```
(setq inferior-lisp-program "sbcl")
```

**Comment by joaotavora** [comment]
Oh, after you switch to the buffer, post its contents here :-)

**Comment by anquegi** [comment]
Hi @joaotavora , thanks for your quick response, first I installed sbcl with brew (I normally use roswell as my lisp distribution manager). That I change that line to your suggestion but same thing happens.

I switch to that buffer and it really pointed the error:

More information about SBCL is available at <http://www.sbcl.org/>.

```
SBCL is free software, provided as is, with absolutely no warranty.
It is mostly in the public domain; some portions are provided under
BSD-style licenses.  See the CREDITS and COPYING files in the
distribution for more information.
* WARNING:
   redefining EMACS-INSPECT (#<SB-PCL:SYSTEM-CLASS COMMON-LISP:T>) in DEFMETHOD

SLYNK's ASDF loader finished.
debugger invoked on a SB-C::INPUT-ERROR-IN-LOAD in thread
#<THREAD "main thread" RUNNING {100197AAF3}>:
  READ error during LOAD:

    Package SWANK does not exist.

      Line: 5, Column: 34, File-Position: 81

      Stream: #<SB-INT:FORM-TRACKING-STREAM for "file /Users/toni/.swank.lisp" {10021F3343}>

Type HELP for debugger help, or (SB-EXT:EXIT) to exit from SBCL.

restarts (invokable by number or by possibly-abbreviated name):
  0: [ABORT                        ] Abort loading file "/Users/toni/.swank.lisp".
  1: [RETRY                        ] Retry
                                     completing load for #<SYSTEM "slynk">.
  2: [ACCEPT                       ] Continue, treating
                                     completing load for #<SYSTEM "slynk"> as
                                     having been successful.
  3:                                 Retry ASDF operation.
  4: [CLEAR-CONFIGURATION-AND-RETRY] Retry ASDF operation after resetting the
                                     configuration.
  5:                                 Exit debugger, returning to top level.

(SB-C:COMPILER-ERROR SB-C::INPUT-ERROR-IN-LOAD :CONDITION #<SB-INT:SIMPLE-READER-PACKAGE-ERROR "Package ~A does not exist." {10021F4F53}> :STREAM #<SB-INT:FORM-TRACKING-STREAM for "file /Users/toni/.swank.lisp" {10021F3343}>)
0] 
```

so I change my initialiation to this:
`
(setq inferior-lisp-program "ros dynamic-space-size=3gb -Q run -- --eval \"(ql:quickload :swank)\"")`

and now it works pretty well, but I know the point of needing swank, but i tshould be loaded by default isn't it?, so now I uninstalled the sbcl (installed with brew)

**Comment by joaotavora** [comment]
Ok great. But now that I understand what happened i believe an easier fix
for would have been too create a (possibly empty) ~\.slynk.lisp or
~\.slynkrc file in your home directory.

To be honest i don't fully understand what you did to fix it but that's
because i don't use roswel.

It's not very important, but what was happening is that SLY tried to load
Slime's ~\.swank.lisp but failed because Swank and Slynk are no longer
compatible. I'll probably change SLY to not do that anymore...

**Comment by anquegi** [comment]
Thanks for your help, I really like [roswell](https://github.com/roswell/roswell) and maybe and integration with roswell that have slime will be good. If I have the level maybe I will try. I will explain what this line do, I really like roswell, is like rvm, jenv or pyenv. The only thing that I do is start the default roswell implementation in this case sbcl, another important thing is that I add 3 gb to the sbcl memory ([it is needed for loading clml](https://github.com/mmaul/clml)), and finally the first thing when sbcl is open I load swank with quicklisp, so the error dissapears. and finally it works

**Comment by joaotavora** [comment]
Well you don't *need* to load swank to fix this issue, but if you *want* to, go ahead :-)

**Comment by amno1** [comment]
Hi guys; I am perceiving a similar issue in a fresh Windows 11 install; Emacs 29.2, whatever latest Sly in Melpa is and sbcl 3.9.2 (latest build that works on Windows). My Emacs setup for windows has been working on my desktop with Window 10 and another laptop with Windows 10, as well as in Arch Linux. However on this new laptop with Windows 11, Sly is just polling without anything happening. sbcl on command line works fine. Any idea? Here is a bit what happends:

Command line:

```
arthu@Emmi MINGW64 ~/repos/sbcl-fasteval
$ src/runtime/sbcl.exe --core ./output/sbcl.core
This is SBCL 2.3.9.76-fbd65addb-WIP, an implementation of ANSI Common Lisp.
More information about SBCL is available at <http://www.sbcl.org/>.

SBCL is free software, provided as is, with absolutely no warranty.
It is mostly in the public domain; some portions are provided under
BSD-style licenses.  See the CREDITS and COPYING files in the
distribution for more information.
*
```
Sly:

```
[sly] Polling "c:/Users/arthu/AppData/Local/Temp/sly.22872" .. 462 (Abort with `M-x sly-abort-connection'.)
[sly] Polling "c:/Users/arthu/AppData/Local/Temp/sly.22872" .. 463 (Abort with `M-x sly-abort-connection'.)
[sly] Polling "c:/Users/arthu/AppData/Local/Temp/sly.22872" .. 464 (Abort with `M-x sly-abort-connection'.)
[sly] Polling "c:/Users/arthu/AppData/Local/Temp/sly.22872" .. 465 (Abort with `M-x sly-abort-connection'.)
[sly] Polling "c:/Users/arthu/AppData/Local/Temp/sly.22872" .. 466 (Abort with `M-x sly-abort-connection'.)
[sly] Polling "c:/Users/arthu/AppData/Local/Temp/sly.22872" .. 467 (Abort with `M-x sly-abort-connection'.)
[sly] Polling "c:/Users/arthu/AppData/Local/Temp/sly.22872" .. 468 (Abort with `M-x sly-abort-connection'.)
[sly] Cancelled connection attempt.
```

I was able to start Slynk server manually from command line and connect to it with sly-connect and specifying the port number.

I can also in the task manager that there is a child sbcl process started, but 0% cpu activity. Every attempt to start sly seems to start a new sbcl child process, but it does not connect.

I made an empty .slynkrc and slynk.lisp in home dir as mentioned in this issue, but that does not help; and I don't have any of those files on my other Windows machines either.

Any idea what happens there? I can't connect to neither sbcl installed into program directory, nor to the one I compiled myself and used on the command line.  Can it be some Windows 11 firewall issue?

**Comment by joaotavora** [comment]
@amno1 please make a separate discussion for this.   It is possibly a Windows issue, not related to the firewall, but to how some Windows executables like (or don't like) to talk over stdio.  At least Allegro CL is known to suffer from this.  https://franz.com/emacs/slime.lhtml  has some interesting notes, as this problem also affects SLIME. This can be worked around with a bat file I remember, though it's not super pretty.  


For Allegro CL (not SBCL) You need to have this `.bat` file:

```
@echo off

if not "%1"=="" goto :start
:noimpl
echo USAGE: start-inferior-lisp.bat c:\path\to\lisp
echo   This command will read a line of lisp forms to start the slynk
echo   server from stdin and start a lisp that runs these forms.
exit 1

:start
set /p slynk_command="Enter the commands necessary to start the slynk server on one line (sly will do this for you)"

echo Starting slynk server in lisp %1...

%1 +B +m -e "%slynk_command%"
```

Then, for a given `foolisp.exe` make a `foolisp.bat` that is just:

```
C:\path\to\aforementioned\start-inferior-lisp.bat C:\path\to\windows\foolisp.exe
```

Then set `inferior-lisp-command` to `foolisp.bat` instead of `foolisp.exe`.  

This is known to work for Allegro CL.  Not sure about SBCL :shrug:

**Comment by amno1** [comment]
Hi Joao; thank you for the pointers and answer.

I have tried it, but it didn't work. I'll take another look tomorrow and will try to see more closely what differentiates the new installation from the old ones. It feels strange it works just fine on two different machines with Win10; but not on the new one. I did nothing special on those; just ordinary msys2 and pacman -S for the stuff I need.

You are correct, it isn't a firewall; otherwise, I wouldn't be able to connect from Sly when I start Slynk from the terminal.

Sorry for reopening this one. Looked to me like a similar issue. Just close it, it is ok. I can make a new issue tomorrow.
